au/code/au/utility/mod.hh

```diff
@@ -0,0 +1,95 @@
+// Copyright 2024 Aurora Operations, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstdint>
+#include <limits>
+
+namespace au {
+namespace detail {
+
+// (a * b) % n
+//
+// Precondition: (a < n).
+// Precondition: (b < n).
+constexpr uint64_t add_mod(uint64_t a, uint64_t b, uint64_t n) {
```

Comment:
`((3 * 2) % 4)` is 2, but `add_mod(3, 2, 4)` is 1. Is that expected?

Reply:
Yes, because `((3 + 2) % 4)` is `1`.  :wink:

(I expect you'll find that `mul_mod(3, 2, 4)` gives `2`.)

Reply:
Ah, I see --- documentation fail.  Let me go fix the comment...

Reply:
Fixed.  Also, thought of a new test case, and found one last overflow bug!  This one was only possible to hit when testing the max int $(2^{64} - 1)$.